src/Pate/MemCell.hs

```diff
@@ -75,121 +73,87 @@ instance PC.TestEquality (WI.SymExpr sym) => Eq (MemCell sym arch w) where
+      Maybe (WI.Pred sym)
+    dropFalse p = case WI.asConstantPred p of
+      Just False -> Nothing
+      _ -> Just p
+
+mergeMemCellPred ::
   WI.IsExprBuilder sym =>
   PC.OrdF (WI.SymExpr sym) =>
   sym ->
-  WI.Pred sym ->
-  MemCells sym arch w ->
-  MemCells sym arch w ->
-  IO (MemCells sym arch w)
-muxMemCells sym p (MemCells cellsT) (MemCells cellsF) = case WI.asConstantPred p of
-  Just True -> return $ MemCells cellsT
-  Just False -> return $ MemCells cellsF
-  _ -> do
-    notp <- WI.notPred sym p
-    MemCells <$>
-      MapM.mergeA
-        (MapM.traverseMissing (\_ pT -> WI.andPred sym pT p))
-        (MapM.traverseMissing (\_ pF -> WI.andPred sym pF notp))
-        (MapM.zipWithAMatched (\_ p1 p2 -> WI.baseTypeIte sym p p1 p2))
-        cellsT
-        cellsF
-
--- | Mux two 'MemCells' maps, where entries that appear in only one map
--- are made conditional on the given predicate.
-muxMemCellsMap ::
+  MemCellPred sym arch ->
+  MemCellPred sym arch ->
+  IO (MemCellPred sym arch)
+mergeMemCellPred sym cells1 cells2 = do
+  MapM.mergeA
+    MapM.preserveMissing
+    MapM.preserveMissing
+    (MapM.zipWithAMatched (\_ p1 p2 -> WI.orPred sym p1 p2))
+    cells1
+    cells2
+
+muxMemCellPred ::
   WI.IsExprBuilder sym =>
   PC.OrdF (WI.SymExpr sym) =>
-  PC.OrdF f =>
   sym ->
   WI.Pred sym ->
-  MapF.MapF f (MemCells sym arch) ->
-  MapF.MapF f (MemCells sym arch) ->
-  IO (MapF.MapF f (MemCells sym arch))
-muxMemCellsMap sym p cellsMapT cellsMapF = case WI.asConstantPred p of
-  Just True -> return cellsMapT
-  Just False -> return cellsMapF
+  MemCellPred sym arch ->
+  MemCellPred sym arch ->
+  IO (MemCellPred sym arch)
+muxMemCellPred sym p cellsT cellsF = case WI.asConstantPred p of
+  Just True -> return cellsT
+  Just False -> return cellsF
   _ -> do
     notp <- WI.notPred sym p
-    MapF.mergeWithKeyM
-         (\_ cellsT cellsF -> Just <$> muxMemCells sym p cellsT cellsF)
-         (TF.traverseF (mapCellPreds (WI.andPred sym p)))
-         (TF.traverseF (mapCellPreds (WI.andPred sym notp)))
-         cellsMapT
-         cellsMapF
-
--- | Unconditionally merge two 'MemCells' maps.
-mergeMemCellsMap ::
-  WI.IsExprBuilder sym =>
-  PC.OrdF (WI.SymExpr sym) =>
-  PC.OrdF f =>
-  sym ->
-  MapF.MapF f (MemCells sym arch) ->
-  MapF.MapF f (MemCells sym arch) ->
-  IO (MapF.MapF f (MemCells sym arch))
-mergeMemCellsMap sym cellsMap1 cellsMap2 = do
-  MapF.mergeWithKeyM
-       (\_ cells1 cells2 -> Just <$> mergeMemCells sym cells1 cells2)
-       return
-       return
-       cellsMap1
-       cellsMap2
+    MapM.mergeA
+      (MapM.traverseMissing (\_ pT -> WI.andPred sym pT p))
+      (MapM.traverseMissing (\_ pF -> WI.andPred sym pF notp))
+      (MapM.zipWithAMatched (\_ p1 p2 -> WI.baseTypeIte sym p p1 p2))
+      cellsT
+      cellsF
 
 -- | True if this cell is logically equivalent to any cell in the given
 -- collection. Note that this is still false if the given cell overlaps
 -- two different entries.
```

Comment:
Could we elaborate this note? Is it saying that it will return `True` if the `MemCell` overlaps just one entry in `MemCellPred`?

Reply:
Elaborated this a bit.